src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java

```diff
@@ -250,109 +307,201 @@ private UserDetails retrieveUser(String username, UsernamePasswordAuthentication
-            }
+                        String dn = dnObject.toString();
+                        LdapName ldapName = new LdapName(dn);
+                        String dnFormatted = ldapName.toString();
+
+                        if (bindName != null && !password.equals(NO_AUTHENTICATION)) {
+                            // if we've used the credential specifically for the bind, we
+                            // need to verify the provided password to do authentication
+                            LOGGER.fine("Attempting to validate password for DN=" + dn);
+                            DirContext test = descriptor.bind(dnFormatted, password, ldapServers);
+                            // Binding alone is not enough to test the credential. Need to actually perform some query operation.
+                            // but if the authentication fails this throws an exception
+                            try {
+                                new LDAPSearchBuilder(test, domainDN).searchOne("(& (userPrincipalName={0})(objectCategory=user))", userPrincipalName);
+                            } finally {
+                                closeQuietly(test);
+                            }
+                        }
 
-            LOGGER.log(Level.WARNING, "Failed to retrieve user information for "+username, e);
-            throw new BadCredentialsException("Failed to retrieve user information for "+username, e);
-        } finally {
-            closeQuietly(context);
+                        Set<GrantedAuthority> groups = resolveGroups(domainDN, dnFormatted, context);
+                        groups.add(SecurityRealm.AUTHENTICATED_AUTHORITY);
+
+                        return new ActiveDirectoryUserDetail(username, password, true, true, true, true, groups.toArray(new GrantedAuthority[groups.size()]),
+                                getStringAttribute(user, "displayName"),
+                                getStringAttribute(user, "mail"),
+                                getStringAttribute(user, "telephoneNumber")
+                        ).updateUserInfo();
+                    } catch (NamingException e) {
+                        if (anonymousBind && e.getMessage().contains("successful bind must be completed") && e.getMessage().contains("000004DC")) {
+                            // sometimes (or always?) anonymous bind itself will succeed but the actual query will fail.
+                            // see JENKINS-12619. On my AD the error code is DSID-0C0906DC
+                            throw new UserMayOrMayNotExistException("Unable to retrieve the user information without bind DN/password configured");
+                        }
+
+                        LOGGER.log(Level.WARNING, "Failed to retrieve user information for " + username, e);
+                        throw new BadCredentialsException("Failed to retrieve user information for " + username, e);
+                    } catch (AuthenticationException e) {
+                        throw e;
```

Comment:
Because later you need to catch it in the next catch `UncheckedExecutionException` to re-throw again so the login is failed

```
catch (UncheckedExecutionException e) {
           Throwable t = e.getCause();
            if (t instanceof AuthenticationException) {
                AuthenticationException authenticationException= (AuthenticationException)t;
                throw authenticationException;
            } else {
                throw new GuavaAuthenticationException("Authentication failed because there was a problem caching user " +  username);
            }
        }
```

Do you suggest a different alternative?